vars/buildDockerAndPublishImage.groovy

```diff
@@ -13,96 +13,73 @@ def call(String imageName, Map config=[:]) {
+              nextVersion = nextVersion + metadata
+            } // if
+          } // container
           echo "Next Release Version = $nextVersion"
-        } // steps
-      } // stage
+        } // stage
+      } // if
+
 
       stage('Prepare') {
-        environment {
-          DOCKER_REGISTRY = credentials("${dockerConfig.credentials}")
-        }
-        steps {
+        withCredentials([usernamePassword(credentialsId: dockerConfig.credentials, passwordVariable: 'DOCKER_REGISTRY_PSW', usernameVariable: 'DOCKER_REGISTRY_USR')]) {
+          // Logging in on the Dockerhub helps to avoid request limit from DockerHub
           sh 'echo "${DOCKER_REGISTRY_PSW}" | img login -u "${DOCKER_REGISTRY_USR}" --password-stdin'
-          script {
-            // Retrieve Makefile
-            writeFile file: 'Makefile', text: makefileContent
-          }
-        } // steps
+
+          // The makefile to use must come from the pipeline to avoid a nasty user trying to exfiltrate data from the build
+          // Even though we have mitigation trhough the multibranch job config only allowed to build PRs from repo. contributors
+          writeFile file: 'Makefile', text: makefileContent
+        } // withCredentials
       } // stage
 
       stage("Lint") {
-        steps {
-          sh 'make lint'
-        } // steps
-        post {
-          always {
-            recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
-          }
-        } // post
-      } //stage
+        int lintStatus = sh(returnStatus: true, script: 'make lint')
+        recordIssues enabledForFailure: true, tool: hadoLint(pattern: 'hadolint.json')
+        if (lintStatus > 0) {
+          currentBuild.result = 'FAILURE'
+          error('Lint Failed')
+        }
+      } // stage
+
 
       stage("Build") {
-        steps {
-          sh 'make build'
-        } // steps
+        sh 'make build'
       } //stage
 
-      stage("Test") {
-        when {
-          expression { fileExists 'cst.yml' }
-        }
-        steps {
-          sh 'make test'
-        } // steps
-      } //stage
 
-      stage("Semantic Release") {
-        when {
-          allOf {
-            expression { env.SEMANTIC_RELEASE == 'true' }
-            branch dockerConfig.mainBranch  
+      // Test step is not mandatory as not all repositories
+      if (fileExists('cst.yml')) {
+        stage("Test") {
+          if (sh(returnStatus: true, script: 'make test') > 0) {
+            currentBuild.result = 'FAILURE'
+            error('Test Failed')
           }
-        }
-        steps {
+        } //stage
+      } // if
+
+      // Automatic tagging on principal branch is not enabled by default
+      if (semVerEnabled) {
+        stage("Semantic Release") {
           echo "Configuring credential.helper"
           sh 'git config --local credential.helper "!f() { echo username=\\$GIT_USERNAME; echo password=\\$GIT_PASSWORD; }; f"'
```

Comment:
Cannot comment on unmodified lines below (thx GitHub) but

```groovy
            echo "Tagging New Version: $nextVersion"
            sh "git tag $nextVersion"
            echo "Pushing Tag"
            sh "git push origin --tags"
```

can be simplified to just

```groovy
            sh 'git tag $nextVersion && git push origin --tags'
```

assuming `nextVersion` is bound as an environment variable (which it should be). Do not ever pass `"`-quoted args to `sh`, and you should never have any series of adjacent `sh` and/or `echo`.

Reply:
Sounds like there is need for a pipeline linter

Reply:
Thanks for the tips @jglick . Gareth and I will check what can be improved on a subsequent PR, as this code block is outside the scope here, but for sure we are willing to improve this to simplify the pipeline code 👍